internal/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,185 @@
+package ratelimit
+
+import (
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+// Ratelimiter holds all ratelimit buckets
+type RateLimiter struct {
+	sync.Mutex
+	buckets         map[string]*Bucket
+	globalRateLimit time.Duration
+}
+
+// New returns a new RateLimiter
+func New() *RateLimiter {
+
+	return &RateLimiter{
+		buckets: make(map[string]*Bucket),
+	}
+}
+
+// getBucket retrieves or creates a bucket
+// ratelimiter is expected to be locked when calling this
+func (r *RateLimiter) getBucket(key string) *Bucket {
+
+	if bucket, ok := r.buckets[key]; ok {
+		return bucket
+	}
+
+	b := &Bucket{remaining: 1, r: r, Key: key}
+	r.buckets[key] = b
+	return b
+}
+
+// LockBucket Locks until a request can be made
+func (r *RateLimiter) LockBucket(path string) *Bucket {
+
+	bucketKey := ParseURL(path)
+
+	r.Lock()
+	b := r.getBucket(bucketKey)
+	r.Unlock()
+
+	b.mu.Lock()
+
+	// If we ran out of calls and the reset time is still ahead of us
+	// then we need to take it easy and relax a little
+	for b.remaining < 1 && b.reset.After(time.Now()) {
```

Comment:
Does this need to be a for loop? It looks like you're sleeping enough to be fine.